sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -3083,6 +3083,17 @@ object SQLConf {
     .booleanConf
     .createWithDefault(false)
 
+  val LEGACY_DATETIME_ENCODER_ENABLED =
+    buildConf("spark.sql.legacy.datetimeEncoder.enabled")
```

Comment:
How about `spark.sql.datetime.java8API.collect.enabled`? This should be a sub-config of `spark.sql.datetime.java8API.enabled`, which enables java8 datetime API everywhere, including data source, UDF and df.collect. It's pretty reasonable to have a sub-config of it to enable the datetime API in a specific place.

Reply:
+1 for the new config name.

Reply:
I wouldn't introduce new **public** config if it is needed only in corner cases. I especially put it into the **legacy** namespace because it is need during the migration new Spark version when a datasource connecter still doesn't support all possible Spark SQL types.